[CSRanking] Allow worse (score-wise) solutions to be removed by filtering

xedin opened the pull request:
The removed condition was incorrect because first of all it would
always be true since `losers` are populated with `false` based on
number of viable solutions and secondly it would result in a mix
of solutions with and without fixes.

Instead, in ambiguity cases, let's remove all of the solutions
that are worse than others based on score to avoid doing any
extra work in the future steps or during diagnostics.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please test source compatibility

LucianoPAlmeida commented on lib/Sema/CSRanking.cpp:
```diff
@@ -1424,63 +1431,15 @@ ConstraintSystem::findBestSolution(SmallVectorImpl<Solution> &viable,
+                 viable.end());
   }
-  viable.erase(viable.begin() + outIndex, viable.end());
-  NumDiscardedSolutions += viable.size() - outIndex;
```
@xedin Just a question, does `NumDiscardedSolutions` still need to be update since some solutions in `viable` are still being erased and if that could have some effect later(not sure where this is used)?

xedin replied:
It's just a perf counter, and I'm not even sure it's a useful one... I'm just trying to see whether this broke anything and then I'm going to deal with that...

xedin replied:
No need to apologize! Please feel free to ask questions when something is unclear :)

xedin committed:
[CSRanking] Allow worse (score-wise) solutions to be removed by filtering

The removed condition was incorrect because first of all it would
always be true since `losers` are populated with `false` based on
number of viable solutions and secondly it would result in a mix
of solutions with and without fixes.

Instead, in ambiguity cases, let's remove all of the solutions
that are worse than others based on score to avoid doing any
extra work in the future steps or during diagnostics.

xedin committed:
[Sema] NFC: Add some unit-tests for solution filtering

xedin commented:
@swift-ci please smoke test
